cmd/acr/purge.go

```diff
@@ -463,43 +467,41 @@ func getManifestsToDelete(ctx context.Context, acrClient api.AcrCLIClientInterfa
 		return nil, err
 	}
 	// This will act as a set if a key is present then it should not be deleted because it is referenced by a multiarch manifest
-	// that will not be deleted
-	doNotDelete := map[string]bool{}
-	candidatesToDelete := []acr.ManifestAttributesBase{}
-	// Iterate over all manifests to discover multiarchitecture manifests
+	// or the manifest has subjects attached
+	doNotDelete := make(Set[string])
+	var candidatesToDelete []acr.ManifestAttributesBase
 	for resultManifests != nil && resultManifests.ManifestsAttributes != nil {
 		manifests := *resultManifests.ManifestsAttributes
 		for _, manifest := range manifests {
-			if (*manifest.MediaType == manifestListContentType || *manifest.MediaType == manifestIndexContentType) && manifest.Tags != nil {
-				// If a manifest list is found and it has tags then all the dependent digests are
-				// marked to not be deleted.
-				var manifestListBytes []byte
-				manifestListBytes, err = acrClient.GetManifest(ctx, repoName, *manifest.Digest)
+			var cm customManifest
+			if _, ok := mediaTypes[*manifest.MediaType]; ok {
+				var manifestBytes []byte
+				manifestBytes, err = acrClient.GetManifest(ctx, repoName, *manifest.Digest)
 				if err != nil {
 					return nil, err
 				}
-				var manifestList multiArchManifest
-				err = json.Unmarshal(manifestListBytes, &manifestList)
+				err = json.Unmarshal(manifestBytes, &cm)
 				if err != nil {
 					return nil, err
 				}
-				for _, dependentDigest := range manifestList.Manifests {
-					doNotDelete[dependentDigest.Digest] = true
-				}
-			} else if manifest.Tags == nil {
-				// If the manifest has no tags left it is a candidate for deletion
-				candidatesToDelete = append(candidatesToDelete, manifest)
+			}
+			candidatesToDelete, doNotDelete, err = filterManifests(manifest, cm, doNotDelete, candidatesToDelete)
```

Comment:
Wondering if it is appropriate to call it `filterManifests`🤔... How about `getCandidateManifestsToDelete`? or `getCandidatesToDelete`?

Reply:
because this method also returns a set of manifests that should not be deleted. we do actually care about both list/set since we later on compare these two to determine the actual manifests to be deleted.

Reply:
discussed offline, using `getCandidatesToDelete` now unless we come up a better name